src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
+
+class AstPredicate {
+  public:
+    AstPredicate(
+        std::string name,
+        std::vector<std::string> args,
+        AstSign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        sign_(std::move(sign)) {}
+  private:
+    std::string name_;
+    std::vector<std::string> args_;
+    AstSign sign_;
+};
+
+// This represents an expression that applies either a predicate or a 
+// "canActAs" to a principal, and it is a sub-term of AstFlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class AstVerbPhrase {
+  public:
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPredicate pred);
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPrincipal right_prin);
+  private:
+    // The principal variant of the internal form means this represents
+    // "canActAs principal".
+    using InternalFormType = std::variant<AstPredicate, AstPrincipal>;
+    InternalFormType internal_form_;
+    AstVerbPhrase(InternalFormType internal_form)
+      : internal_form_(std::move(internal_form)) {}
```

Comment:
Although I am sympathetic to this use of an alias to ensure that the same type is used for the field and the constructor parameter, I believe it is a violation of the rule [not to use aliases in the public API](https://google.github.io/styleguide/cppguide.html#Aliases).

In addition, I believe that, under style guide declaration order rules, the [constructor must come before the field](https://google.github.io/styleguide/cppguide.html#Aliases).

Reply:
Why not create public constructors for the class?

```C++
explicit AstVerbPhrase(AstPredicate predicate) {...}
explicit AstVerbPhrase(AstPrincipal predicate) {...}
```

Reply:
I have no idea yet if using these variants to write more substantial code becomes unwieldy, though. I guess I/we'll find out shortly.

Reply:
Regarding my comment about constructors what I meant was define constructors as follows:

```C++
class VerbPhrase {
 public:
   explicit VerbPhrase(Predicate pred): internal_form_(pred) {}
   explicit VerbPhrase(Principal right_prin): internal_form_(right_prin) {}
 private:
  // The principal variant of the internal form means this represents
  // "canActAs principal".
  std::variant<Predicate, Principal> internal_form_;
};
```